File: packages/core/src/schemas/metadata/MetadataTemplate.ts
import {
  ClassProperties,
  IMetadataSchema,
  IMetadataTypeTag,
} from "@typia/interface";

import { IMetadataDictionary } from "./IMetadataDictionary";
import { MetadataSchema } from "./MetadataSchema";

export class MetadataTemplate {
  public readonly row: MetadataSchema[];
  public readonly tags: IMetadataTypeTag[][];

  private name_?: string;

  private constructor(props: ClassProperties<MetadataTemplate>) {
    this.row = props.row.map(MetadataSchema.create);
    this.tags = props.tags;
  }

  /** @internal */
  public static create(
    props: ClassProperties<MetadataTemplate>,
  ): MetadataTemplate {
    return new MetadataTemplate(props);
  }

  public static from(
    json: IMetadataSchema.ITemplate,
    dict: IMetadataDictionary,
  ): MetadataTemplate {
    return new MetadataTemplate({
      row: json.row.map((m) => MetadataSchema.from(m, dict)),
      tags: json.tags,
    });
  }

  public getName(): string {
    return (this.name_ ??= getName(this));
  }

  /** @internal */
  public getBaseName(): string {
    return (
      "`" +
      this.row
        .map((child) =>
          child.isConstant() && child.size() === 1
            ? child.constants[0]!.values[0]!.value
            : `$\{${child.getName()}\}`,
        )
        .join("")
        .split("`")
        .join("\\`") +
      "`"
    );
  }

  public toJSON(): IMetadataSchema.ITemplate {
    return {
      row: this.row.map((m) => m.toJSON()),
      tags: this.tags,
    };
  }
}

const getName = (template: MetadataTemplate): string => {
  const base: string = template.getBaseName();
  if (!template.tags?.length) return base;
  else if (template.tags.length === 1) {
    const str: string = [base, ...template.tags[0]!.map((t) => t.name)].join(
      " & ",
    );
    return `(${str})`;
  }
  const rows: string[] = template.tags.map((row) => {
    const str: string = row.map((t) => t.name).join(" & ");
    return row.length === 1 ? str : `(${str})`;
  });
  return `(${base} & (${rows.join(" | ")}))`;
};

File: packages/core/src/schemas/metadata/MetadataTemplate.go
package metadata

import (
	"fmt"
	"strings"
)

type MetadataTemplate struct {
	Row  []*MetadataSchema
	Tags [][]MetadataTypeTag

	name string
}

type MetadataTemplateProps struct {
	Row  []*MetadataSchema
	Tags [][]MetadataTypeTag
}

func CreateMetadataTemplate(props MetadataTemplateProps) *MetadataTemplate {
	return &MetadataTemplate{
		Row:  append([]*MetadataSchema(nil), props.Row...),
		Tags: cloneTagMatrix(props.Tags),
	}
}

func MetadataTemplateFrom(json MetadataTemplateJSON, dict *IMetadataDictionary) (*MetadataTemplate, error) {
	row := make([]*MetadataSchema, 0, len(json.Row))
	for _, elem := range json.Row {
		next, err := MetadataSchemaFrom(elem, dict)
		if err != nil {
			return nil, err
		}
		row = append(row, next)
	}
	return CreateMetadataTemplate(MetadataTemplateProps{
		Row:  row,
		Tags: json.Tags,
	}), nil
}

func (m *MetadataTemplate) GetName() string {
	if m.name == "" {
		m.name = taggedReferenceName(m.GetBaseName(), m.Tags)
	}
	return m.name
}

func (m *MetadataTemplate) GetBaseName() string {
	parts := make([]string, 0, len(m.Row))
	for _, child := range m.Row {
		if child.IsConstant() && child.Size() == 1 && len(child.Constants) == 1 && len(child.Constants[0].Values) == 1 {
			parts = append(parts, fmt.Sprint(child.Constants[0].Values[0].Value))
		} else {
			parts = append(parts, "${"+child.GetName()+"}")
		}
	}
	return "`" + strings.ReplaceAll(strings.Join(parts, ""), "`", "\\`") + "`"
}

func (m *MetadataTemplate) ToJSON() MetadataTemplateJSON {
	row := make([]MetadataSchemaJSON, 0, len(m.Row))
	for _, elem := range m.Row {
		row = append(row, elem.ToJSON())
	}
	return MetadataTemplateJSON{
		Row:  row,
		Tags: cloneTagMatrix(m.Tags),
	}
}
